### PR TITLE
SPU liveness check

### PR DESCRIPTION
**Pull request by nacardin**
Resolves https://github.com/infinyon/fluvio/issues/849

### CODE REVIEW

File: src/client/src/error.rs [hunk 1]
```diff
@@ -7,12 +7,14 @@ use crate::config::ConfigError;
 use semver::Version;
 
 /// Possible errors that may arise when using Fluvio
+#[allow(clippy::upper_case_acronyms)]
```
Comment (nicholastmosher) [inline src/client/src/error.rs:10]: We added `-A clippy::upper_case_acronyms` to the `make check-clippy` task, so I think you shouldn't need to have these to pass CI now

File: src/extension-runner-local/src/bin/main.rs [hunk 2]
```diff
@@ -14,8 +14,11 @@ fn main() {
 #[derive(Debug, StructOpt)]
 pub enum RunnerCmd {
     /// Run a Streaming Controller (SC) or SPU
-    #[structopt(name = "run")]
+    #[structopt()]
     Run(RunCmd),
+    /// Run a new Streaming Processing Unit (SPU)
+    #[structopt()]
```
Comment (nicholastmosher) [inline src/extension-runner-local/src/bin/main.rs:20]: Is this supposed to have a short name?

**Comment by nicholastmosher** [comment]
I see there's a new CLI command here, what is the intended usage? The command is on the SPU cluster component, so is it supposed to be a different process that also runs inside the cluster? Is this code the client or the server to probe requests?

File: src/client/src/error.rs (hunk 1 above)
Comment (sehz) [reply src/client/src/error.rs:10]: I agree.

File: src/sc/src/k8/objects/spg_group.rs [hunk 3]
```diff
@@ -217,16 +220,29 @@ mod k8_convert {
...
+            args.push(format!("0.0.0.0:{}", spu_non_tls_public));
         }
 
         // add RUST LOG, if passed
         if let Ok(rust_log) = std::env::var("RUST_LOG") {
             env.push(Env::key_value("RUST_LOG", &rust_log));
         }
 
-        // env.append(&mut spu_template.env.clone());
+        let liveness_probe = Probe {
```
Comment (sehz) [inline src/sc/src/k8/objects/spg_group.rs:234]: Don't think liveness probes should run in the SC

File: src/spu/src/liveness_check.rs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    net::{SocketAddr, TcpStream},
+};
+use std::io::prelude::*;
+
+use tracing::debug;
+
+use fluvio_protocol::{Encoder, api::RequestMessage};
+use dataplane::versions::ApiVersionsRequest;
+
+pub fn probe(endpoint: SocketAddr) -> Result<()> {
```
Comment (sehz) [inline src/spu/src/liveness_check.rs:13]: Is checking SPU or SC?

File: src/client/src/error.rs (hunk 1 above)
Comment (nacardin) [reply src/client/src/error.rs]: Removed

**Review by sehz — CHANGES_REQUESTED** [review]

File: src/extension-runner-local/src/bin/main.rs (hunk 2 above)
Comment (nacardin) [reply src/extension-runner-local/src/bin/main.rs:20]: No

File: src/spu/src/liveness_check.rs (hunk 4 above)
Comment (nacardin) [reply src/spu/src/liveness_check.rs:13]: SPU

File: src/sc/src/k8/objects/spg_group.rs (hunk 3 above)
Comment (nacardin) [reply src/sc/src/k8/objects/spg_group.rs:234]: Here it is being added to the SPG StatefulSet config

**Comment by nacardin** [comment]
> I see there's a new CLI command here, what is the intended usage? The command is on the SPU cluster component, so is it supposed to be a different process that also runs inside the cluster? Is this code the client or the server to probe requests?

K8 executes this as a separate process on an interval in order to check for liveness. This is run inside the same pod as the SPU. This is a client which checks that the SPU is responsive.

**Comment by sehz** [comment]
bors +r

File: src/sc/src/k8/objects/spg_group.rs (hunk 3 above)
Comment (nicholastmosher) [reply src/sc/src/k8/objects/spg_group.rs:234]: This just allows the SC to configure the K8 probe right? So K8 will be running the probe but the SC is the thing that sets it up in the  cluster? If I am interpreting the flow right

Comment (nacardin) [reply src/sc/src/k8/objects/spg_group.rs:234]: @nicholastmosher Correct

File: src/spu/src/liveness_check.rs (hunk 4 above)
Comment (sehz) [reply src/spu/src/liveness_check.rs:13]: As I understand, probe should exist outside application not inside

Comment (nicholastmosher) [reply src/spu/src/liveness_check.rs:13]: I think what he's doing is adding the client and server side of this probe in the same executable. So the SPU server running in the pod will respond to probe requests, but the K8 cluster also needs to know how to send requests that the server will respond to. So this subcommand of the SPU executable will act as a client for that probe request. Is that right @nacardin?

Comment (nacardin) [reply src/spu/src/liveness_check.rs:13]: @nicholastmosher Correct

File: src/extension-runner-local/src/bin/main.rs (hunk 2 above)
Comment (nicholastmosher) [reply src/extension-runner-local/src/bin/main.rs:20]: Gotcha. Ok, then this attribute isn't really necessary :P but it also doesn't hurt really

**Comment by nicholastmosher** [comment]
How can I observe this at work? Is there a way to watch the probe going?

**Comment by nacardin** [comment]
> How can I observe this at work? Is there a way to watch the probe going?

Just `fluvio cluster start --develop`

**Pull request closed without merging**